src/frequenz/sdk/actor/power_distributing/_component_status.py

```diff
@@ -0,0 +1,109 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Classes to track the status of components in the microgrid."""
+
+
+import enum
+import typing
+from abc import ABC, abstractmethod
+from collections import abc
+from dataclasses import dataclass
+
+from frequenz.channels import Receiver, Sender
+
+
+@dataclass
+class ComponentPoolStatus:
+    """Status of all components of a certain category in the microgrid."""
+
+    working: set[int]
+    """Set of working component ids."""
+
+    uncertain: set[int]
+    """Set of components to be used only when there are none known to be working."""
+
+    def get_working_components(self, components: abc.Set[int]) -> set[int]:
+        """From the given set of components return the working ones.
+
+        Args:
+            components: Set of components.
+
+        Returns:
+            Subset with working components.
+        """
+        working = self.working.intersection(components)
+        if len(working) > 0:
+            return working
+        return self.uncertain.intersection(components)
+
+
+class ComponentStatusEnum(enum.Enum):
+    """Enum for component status."""
+
+    NOT_WORKING = 0
+    """Component is not working and should not be used."""
+
+    UNCERTAIN = 1
+    """Component should work, although the last request to it failed.
+
+    It is blocked for few seconds and it is not recommended to use it unless it is
+    necessary.
+    """
+
+    WORKING = 2
+    """Component is working"""
+
+
+@dataclass(frozen=True)
+class ComponentStatus:
+    """Status of a single component."""
+
+    component_id: int
+    """Component ID."""
+
+    value: ComponentStatusEnum
+    """Component status."""
+
```

Comment:
I'm not convinced about this change. Can't you just use `select()` if you need to know the source?  I guess you'll still need a mapping between receiver and `component_id`, but otherwise this is mostly doing the same, but you need to create one extra class that has a name clash with the `Enum`, so now you need to add the `Enum` suffix to the enum, which is weird and we never do (it could be like adding a `Class` suffix to classes.

If you think it is better to stick with `merge` and this mapping wrapper, then I suggest to find some other name to avoid the `Enum` suffix, maybe `ComponentStatus` and `ComponentState`, or `CurrentComponentState` to avoid confusion between status and state?

Reply:
We need the component id from a variable number of components, so implementation-wise I think this is the best approach.

I think they are both status or both state, but one of them is tied to a specific component while the other is not.  I agree that `Enum` is not perfect, but I couldn't come up with a better one.

I would like the class with the id to be called `ComponentStatus`, to form a pattern with the `ComponentPoolStatus`, `ComponentStatusTracker`, `ComponentPoolStatusTracker`.  I guess we just need an alternative for the enum.

Reply:
> We need the component id from a variable number of components, so implementation-wise I think this is the best approach.

just to contextualize this, this is the case if we want to limit the changes to this part of the code base.  it would be possible to avoid the merge if we rewrite the `ComponentPoolStatusTracker`, but that won't be a trivial change.

Reply:
> I would like the class with the id to be called `ComponentStatus`, to form a pattern with the `ComponentPoolStatus`, `ComponentStatusTracker`, `ComponentPoolStatusTracker`. I guess we just need an alternative for the enum.

I see. I think, besides being a bit ambiguous, `ComponentState` for the enum makes sense. Status and state is not exactly the same. For example a state machine is never called status machine, and "WORKING", etc. look like states in a state machine, is like an abstract concept, while status conveys more the meaning of something that's happening at a particular point in time. For example a status report wouldn't be called state report, and it talks about one point in time. `ComponentStatus` says this particular component is like this right now.

Anyway, I agree it has the downside of being a bit ambiguous. Not a perfect solution so I would leave it to you to decide.

Reply:
I was thinking that another option, and in general for all enums, it might make sense to name them in plural, as it is a collection of names or stuff, specially in Python that they are even iterable.